src/Server/TextDocument.php

```diff
@@ -124,4 +127,25 @@ private function updateAst(string $uri, string $content)
             $this->asts[$uri] = $stmts;
         }
     }
+
+    /**
+     * The document formatting request is sent from the server to the client to format a whole document.
+     *
+     * @param TextDocumentIdentifier $textDocument The document to format
+     * @param FormattingOptions $options The format options
+     * @return TextEdit[]
+     */
+    public function formatting(TextDocumentIdentifier $textDocument, FormattingOptions $options)
+    {
+        $nodes = $this->asts[$textDocument->uri];
+        if (empty($nodes)) {
+            return [];
+        }
+        $prettyPrinter = new PrettyPrinter();
+        $edit = new TextEdit();
+        $edit->range = new Range(new Position(0, 0), new Position(PHP_INT_MAX, PHP_INT_MAX));
```

Comment:
@felixfbecker yes and no. It is always best if an extension provides a minimal edit set since it will preserve the most markers attached to the text. We have code in place that runs a diff in the whole document gets replaced to compute some more minimal diff. However this might be suboptimal as well. So the more fine grain your edits are the better.